![image](competi.png)

#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [2]:
import numpy as np
import pandas as pd

import urllib.request
from PIL import Image
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from catboost import Pool, CatBoostRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import LabelEncoder
import seaborn as sns
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler

In [8]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [6]:
df

carat        cut color clarity  depth  table     x     y     z  price
0       1.21      Ideal     H     VS2   63.0   57.0  6.73  6.70  4.23   6134
1       0.28  Very Good     D    VVS2   64.0   56.0  4.14  4.17  2.66    532
2       0.42    Premium     F     VS1   61.2   58.0  4.86  4.82  2.96   1103
3       0.26      Ideal     H      IF   61.1   57.0  4.16  4.12  2.53    600
4       1.10       Good     G     SI1   63.4   57.0  6.52  6.55  4.14   4997
...      ...        ...   ...     ...    ...    ...   ...   ...   ...    ...
40340   1.55    Premium     H     VS2   61.3   61.0  7.46  7.39  4.55  11708
40341   0.36      Ideal     D     SI1   60.6   56.0  4.58  4.63  2.79    619
40342   0.57  Very Good     I     VS2   62.2   55.0  5.33  5.34  3.32   1267
40343   1.01  Very Good     F      IF   59.6   62.0  6.47  6.56  3.88   9965
40344   0.54      Ideal     E     SI2   60.4   57.0  5.33  5.27  3.20   1340

[40345 rows x 10 columns]

In [9]:
df = pd.get_dummies(df)
df

carat  depth  table     x     y     z  price  cut_Fair  cut_Good  \
0       1.21   63.0   57.0  6.73  6.70  4.23   6134         0         0   
1       0.28   64.0   56.0  4.14  4.17  2.66    532         0         0   
2       0.42   61.2   58.0  4.86  4.82  2.96   1103         0         0   
3       0.26   61.1   57.0  4.16  4.12  2.53    600         0         0   
4       1.10   63.4   57.0  6.52  6.55  4.14   4997         0         1   
...      ...    ...    ...   ...   ...   ...    ...       ...       ...   
40340   1.55   61.3   61.0  7.46  7.39  4.55  11708         0         0   
40341   0.36   60.6   56.0  4.58  4.63  2.79    619         0         0   
40342   0.57   62.2   55.0  5.33  5.34  3.32   1267         0         0   
40343   1.01   59.6   62.0  6.47  6.56  3.88   9965         0         0   
40344   0.54   60.4   57.0  5.33  5.27  3.20   1340         0         0   

       cut_Ideal  ...  color_I  color_J  clarity_I1  clarity_IF  clarity_SI1  \
0              1  ...        0        0           0           0            0   
1              0  ...        0        0           0           0            0   
2              0  ...        0        0           0           0            0   
3              1  ...        0        0           0           1            0   
4              0  ...        0        0           0           0            1   
...          ...  ...      ...      ...         ...         ...          ...   
40340          0  ...        0        0           0           0            0   
40341          1  ...        0        0           0           0            1   
40342          0  ...        1        0           0           0            0   
40343          0  ...        0        0           0           1            0   
40344          1  ...        0        0           0           0            0   

       clarity_SI2  clarity_VS1  clarity_VS2  clarity_VVS1  clarity_VVS2  
0                0            0            1             0             0  
1                0            0            0             0             1  
2                0            1            0             0             0  
3                0            0            0             0             0  
4                0            0            0             0             0  
...            ...          ...          ...           ...           ...  
40340            0            0            1             0             0  
40341            0            0            0             0             0  
40342            0            0            1             0             0  
40343            0            0            0             0             0  
40344            1            0            0             0             0  

[40345 rows x 27 columns]

In [3]:
#lb_make = LabelEncoder()
#df['cut'] = lb_make.fit_transform(df['cut'])
#df['color'] = lb_make.fit_transform(df['color'])
#df['clarity'] = lb_make.fit_transform(df['clarity'])

In [242]:
#onehotencoder = OneHotEncoder(sparse=False, handle_unknown="error", drop="first")
#ohe = pd.DataFrame(onehotencoder.fit_transform(df[["color"]]))
#df = df.join(ohe)
#df.rename(columns={0: "a", 1: "b", 2: "c", 3: "d", 4: "e", 5: "f"}, inplace=True)
#oht = pd.DataFrame(onehotencoder.fit_transform(df[["clarity"]]))
#df = df.join(oht)
#df.rename(columns={0: "g", 1: "h", 2: "i", 3: "j", 4: "k", 5: "m", 6: "o", 7: "o", 8: "p"}, inplace=True)
#df["carat"] = df["carat"]*100
#df["price"] = df["price"]/100
#df["x"] = df["x"]*9
#df["y"] = df["y"]*9
#df["z"] = df["z"]*9
df["x"] = df["x"] + df["y"] + df["z"]

In [243]:
column = 'carat'
#df["carat"] = MinMaxScaler().fit_transform(np.array(df[column]).reshape(-1,1))
#df["carat"] = df["carat"]*100
df

carat        cut color clarity  depth  table      x     y     z  price
0       1.21      Ideal     H     VS2   63.0   57.0  17.66  6.70  4.23   6134
1       0.28  Very Good     D    VVS2   64.0   56.0  10.97  4.17  2.66    532
2       0.42    Premium     F     VS1   61.2   58.0  12.64  4.82  2.96   1103
3       0.26      Ideal     H      IF   61.1   57.0  10.81  4.12  2.53    600
4       1.10       Good     G     SI1   63.4   57.0  17.21  6.55  4.14   4997
...      ...        ...   ...     ...    ...    ...    ...   ...   ...    ...
40340   1.55    Premium     H     VS2   61.3   61.0  19.40  7.39  4.55  11708
40341   0.36      Ideal     D     SI1   60.6   56.0  12.00  4.63  2.79    619
40342   0.57  Very Good     I     VS2   62.2   55.0  13.99  5.34  3.32   1267
40343   1.01  Very Good     F      IF   59.6   62.0  16.91  6.56  3.88   9965
40344   0.54      Ideal     E     SI2   60.4   57.0  13.80  5.27  3.20   1340

[40345 rows x 10 columns]

**Importante:**

   - Si quitas columnas o creas columnas nuevas a partir de otras, o cualquier modificación column-wise tendrás que aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción.
   - Si por lo contrario, decides por ejemplo, quitar los outliers o hacer un `dropna()`, o cualquier modificación row-wise eso NO PODRÁS (ni debes) aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción. ¿Por qué? Porque si el conjunto de test tiene 50 observaciones (filas) la predicción se espera que tenga 50 filas.

#### ¿Y si lo que hago es estandarizar los datos o hacer un encoding, también se lo tengo que hacer al test antes de predecir?

-----------------------------------------------------------------------------------------------------------------

### Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [55]:
X = np.array(df[['carat','depth', 'table', 'x', 'y', 'z', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2']])

In [35]:
y = np.array(df["price"])

In [5]:
# coger las columnas "x", "y", "z" tiene solo valor ilustrativo, NO LO COPIÉIS
X = np.array(df[["carat", "cut",  "color", "clarity", "depth",  "table", "x", "y", "z"]]) #without cut, depth
y = np.array(df["price"])
X.shape

(40345, 9)

In [6]:
y.shape

(40345,)

#### ¿Y si después del análisis exploratorio de mis datos llego a la conclusión de que puedo predecir el precio solo con las columnas `x`, `y` y `z`, también aplica al test?

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [ ]:
# ¿qué opináis?
# sí, el conjunto para predecir `diamonds_test.csv` también tendrá que tener 3 columnas

In [10]:
train_data = X_train
train_label = y_train
test_dataX = X_test
test_datay = y_test


train_pool = Pool(train_data, 
                  train_label,
                  )
test_pool = Pool(test_dataX,
                 test_datay)

In [103]:
model = CatBoostRegressor()   #bootstrap_type=Poisson, Bayesian, Bernoulli, MVS

In [104]:
model.fit(train_pool)

0:	learn: 3297.9260486	total: 7.44ms	remaining: 737ms
1:	learn: 2735.6033486	total: 12.4ms	remaining: 610ms
2:	learn: 2300.7203608	total: 16.8ms	remaining: 544ms
3:	learn: 1961.5449132	total: 22.5ms	remaining: 539ms
4:	learn: 1703.1151325	total: 27.5ms	remaining: 522ms
5:	learn: 1484.4805242	total: 32.1ms	remaining: 504ms
6:	learn: 1322.2915034	total: 37.3ms	remaining: 495ms
7:	learn: 1194.5149848	total: 42.6ms	remaining: 489ms
8:	learn: 1094.9325868	total: 47.5ms	remaining: 480ms
9:	learn: 1020.8052717	total: 53ms	remaining: 477ms
10:	learn: 952.9394163	total: 58.5ms	remaining: 473ms
11:	learn: 903.6390374	total: 63.6ms	remaining: 466ms
12:	learn: 863.5927545	total: 68.6ms	remaining: 459ms
13:	learn: 821.8416675	total: 72.9ms	remaining: 448ms
14:	learn: 794.8111309	total: 77ms	remaining: 436ms
15:	learn: 773.7690407	total: 81.1ms	remaining: 426ms
16:	learn: 748.0523261	total: 86.5ms	remaining: 422ms
17:	learn: 734.0196808	total: 91.3ms	remaining: 416ms
18:	learn: 715.7603360	total: 95

In [105]:
preds = model.predict(test_pool)

In [107]:
mean_absolute_error(test_datay, preds)

311.40743425153136

In [54]:
model2 = CatBoostRegressor(learning_rate=0.05, depth=6, l2_leaf_reg=1, iterations=150)

grid = {'learning_rate': [0.05],
        'depth': [6],
        'l2_leaf_reg': [1]}

grid_search_result = model2.grid_search(grid, 
                                       X, 
                                       y, 
                                       plot=True) #cat_features=[1, 2, 3]

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 5061.3676734	test: 5137.9449398	best: 5137.9449398 (0)	total: 13.1ms	remaining: 13.1s
1:	learn: 4603.7200875	test: 4672.8195566	best: 4672.8195566 (1)	total: 17.8ms	remaining: 8.87s
2:	learn: 4201.0213823	test: 4262.7547089	best: 4262.7547089 (2)	total: 22.1ms	remaining: 7.33s
3:	learn: 3841.4085964	test: 3896.5602084	best: 3896.5602084 (3)	total: 27.3ms	remaining: 6.79s
4:	learn: 3520.6534218	test: 3569.4695175	best: 3569.4695175 (4)	total: 32.1ms	remaining: 6.4s
5:	learn: 3233.5190613	test: 3276.1775763	best: 3276.1775763 (5)	total: 36.7ms	remaining: 6.07s
6:	learn: 2977.0900986	test: 3014.1806783	best: 3014.1806783 (6)	total: 42.8ms	remaining: 6.07s
7:	learn: 2748.0984591	test: 2780.1113410	best: 2780.1113410 (7)	total: 47.5ms	remaining: 5.89s
8:	learn: 2550.5284391	test: 2578.4410538	best: 2578.4410538 (8)	total: 52.1ms	remaining: 5.74s
9:	learn: 2378.6573777	test: 2403.3456667	best: 2403.3456667 (9)	total: 56.7ms	remaining: 5.62s
10:	learn: 2220.5686587	test: 2242.686593

In [56]:
model2 = CatBoostRegressor()

grid = {'learning_rate': [0.1],
        'depth': [6],
        'l2_leaf_reg': [1]}

randomized_search_result = model2.randomized_search(grid,
                                                   X,
                                                   y,
                                                   plot=True)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

0:	learn: 5056.0854715	test: 5135.3043220	best: 5135.3043220 (0)	total: 9.51ms	remaining: 9.5s
1:	learn: 4602.6753428	test: 4671.6950045	best: 4671.6950045 (1)	total: 14.9ms	remaining: 7.45s
2:	learn: 4188.2484047	test: 4249.4956554	best: 4249.4956554 (2)	total: 19.6ms	remaining: 6.51s
3:	learn: 3823.0885473	test: 3877.1838767	best: 3877.1838767 (3)	total: 24.1ms	remaining: 6s
4:	learn: 3493.4861776	test: 3539.9540463	best: 3539.9540463 (4)	total: 29.3ms	remaining: 5.82s
5:	learn: 3201.9168565	test: 3242.5845640	best: 3242.5845640 (5)	total: 34.1ms	remaining: 5.65s
6:	learn: 2948.2566133	test: 2984.1014344	best: 2984.1014344 (6)	total: 38.7ms	remaining: 5.49s
7:	learn: 2717.2865172	test: 2749.0183658	best: 2749.0183658 (7)	total: 43.6ms	remaining: 5.4s
8:	learn: 2511.5953350	test: 2538.3969431	best: 2538.3969431 (8)	total: 48.4ms	remaining: 5.33s
9:	learn: 2321.3796591	test: 2344.8393940	best: 2344.8393940 (9)	total: 53ms	remaining: 5.25s
10:	learn: 2158.4821474	test: 2179.9293871	best

### 2. Dividir X_train, X_test, y_train, y_test

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

Siempre hay tiempo para una historia:
https://catboost.ai/docs/concepts/python-reference_catboostregressor.html

### 4. Entrenar el modelo con X_train, y_train

### 5. Predecir con el modelo ya entrenado con X_test

In [57]:
predictions = model2.predict(X)
print(predictions)

[6242.77645251  602.88297808 1126.88358341 ... 1297.45915921 9824.22284218
 1324.02834571]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

In [58]:
np.sqrt(mean_squared_error(y, predictions))

398.8991194939071

In [59]:
mean_absolute_error(y, predictions)

228.64342406277927

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [40]:
X_pred = pd.read_csv("diamonds_test.csv", index_col = 0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

In [17]:
X_pred.shape

(13449, 9)

In [41]:
X_pred = pd.get_dummies(X_pred)

In [18]:
lb_make = LabelEncoder()
X_pred['cut'] = lb_make.fit_transform(X_pred['cut'])
X_pred['color'] = lb_make.fit_transform(X_pred['color'])
X_pred['clarity'] = lb_make.fit_transform(X_pred['clarity'])
#X_pred["table"] = MinMaxScaler().fit_transform(np.array(X_pred[column]).reshape(-1,1))
X_pred

carat  cut  color  clarity  depth  table     x     y     z
0       0.30    2      4        3   60.0   56.0  4.41  4.43  2.65
1       0.34    2      0        1   62.1   57.0  4.52  4.46  2.79
2       1.57    4      5        5   60.3   58.0  7.58  7.55  4.56
3       0.31    2      4        5   61.8   57.0  4.32  4.36  2.68
4       1.51    1      5        6   64.0   60.0  7.26  7.21  4.63
...      ...  ...    ...      ...    ...    ...   ...   ...   ...
13444   1.05    2      2        3   60.3   55.0  6.55  6.68  3.99
13445   0.27    4      3        4   61.9   56.0  4.13  4.16  2.56
13446   1.50    3      0        5   61.8   60.0  7.30  7.37  4.53
13447   1.70    3      2        3   60.2   61.0  7.79  7.68  4.66
13448   0.30    2      0        5   61.2   57.0  4.29  4.34  2.64

[13449 rows x 9 columns]

In [42]:
X_pred = np.array(X_pred[['carat','depth', 'table', 'x', 'y', 'z', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2']])

In [43]:
X_pred.shape

(13449, 26)

**¿Por qué da error?** 

In [44]:
predictions_submit = model2.predict(X_pred)
predictions_submit

array([  424.22915443,  2347.16900377,  9767.15631798, ...,
       14339.94875975, 11107.35575409,   705.39105122])

In [45]:
predictions_submit.shape

(13449,)

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**¿Qué más habrá que quitar o hacer?**

In [ ]:
# ¿Qué opináis?
# En este caso, quitar todas las columnas, menos la x, y y z

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**Truqui**: Carga X que debe estar actualmente actualizada a cómo la usaste para entrenar a `model` y haz que `X_pred` sea igual. ¡Sin quitar filas!

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

In [99]:
#X_pred

array([[4.41, 4.43, 2.65],
       [4.52, 4.46, 2.79],
       [7.58, 7.55, 4.56],
       ...,
       [7.3 , 7.37, 4.53],
       [7.79, 7.68, 4.66],
       [4.29, 4.34, 2.64]])

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

In [47]:
#predictions_submit = model.predict(X_pred)
#predictions_submit

array([   18.68779645,  2500.80852261,  4864.54195375, ...,
        6118.55288359, 13395.34483755,  1887.97613812])

In [48]:
#predictions_submit.shape

(13449,)

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `sample_submission.csv`?**

In [46]:
sample = pd.read_csv("sample_submission.csv")

In [47]:
sample.head()

id  price
0   0  12132
1   1  11786
2   2  14684
3   3  15425
4   4   6724

In [48]:
sample.shape

(13449, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [49]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

In [50]:
submission.head()

id        price
0   0   424.229154
1   1  2347.169004
2   2  9767.156318
3   3   509.484266
4   4  9691.155253

In [51]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [30]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [58]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [52]:
chequeator(submission)

You're ready to submit!


URLError: <urlopen error [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond>